helpers/stable_swap_math/mod.rs

```diff
@@ -0,0 +1,934 @@
+                .try_into()
+                .map_err(|_| MathError::CastOverflow(3))?;
+
+            let diff_shares: u128 = U256::from(pool_token_supply)
+                .checked_mul(d_1.checked_sub(d_0).ok_or(MathError::SubUnderflow(20))?)
+                .ok_or(MathError::MulOverflow(19))?
+                .checked_div(d_0)
+                .ok_or(MathError::DivByZero(11))?
+                .try_into()
+                .map_err(|_| MathError::CastOverflow(4))?;
+            // d1 > d2 > d0,
+            // (d2-d0) => mint_shares (charged fee),
+            // (d1-d0) => diff_shares (without fee),
+            // (d1-d2) => fee part,
+            // diff_shares = mint_shares + fee part
+            Ok((
+                mint_shares,
+                diff_shares
+                    .checked_sub(mint_shares)
+                    .ok_or(MathError::SubUnderflow(21))?,
+            ))
+        } else {
+            // Calc without fees
+            let mint_shares: u128 = U256::from(pool_token_supply)
+                .checked_mul(d_1.checked_sub(d_0).ok_or(MathError::SubUnderflow(22))?)
+                .ok_or(MathError::MulOverflow(20))?
+                .checked_div(d_0)
+                .ok_or(MathError::DivByZero(12))?
+                .try_into()
+                .map_err(|_| MathError::CastOverflow(5))?;
+            // d1 > d0,
+            // (d1-d0) => mint_shares
+            Ok((mint_shares, 0))
+        }
+    }
+}
+
+pub fn rated_compute_lp_amount_for_deposit(
+    rates: &[u128],
+    deposit_amounts: &Vec<u128>,
+    old_reserves: &Vec<u128>,
+    pool_token_supply: u128,
+    fees: Option<&Fees>,
+    amp_coef: u128,
+) -> Result<(u128, u128), MathError> {
+    let r_deposit_amounts = amounts_to_rated(deposit_amounts, rates)?;
+    let r_old_reserves = amounts_to_rated(old_reserves, rates)?;
+
+    compute_lp_amount_for_deposit(
+        &r_deposit_amounts,
+        &r_old_reserves,
+        pool_token_supply,
+        fees,
+        amp_coef,
+    )
+}
+
+/// Compute the ideal amounts of deposits for lp mint
+/// return <deposit_amounts, new_reserves>
+pub fn compute_deposit_amounts_for_lp(
+    lp_amount: u128,
+    old_reserves: &Vec<u128>,
+    pool_token_supply: u128,
+) -> Result<(Vec<u128>, Vec<u128>), MathError> {
+    let mut amounts = Vec::new();
+    let mut new_reserves = Vec::new();
+    for i in 0..old_reserves.len() {
+        amounts.push(
+            U256::from(old_reserves[i])
+                .checked_mul(lp_amount.into())
+                .ok_or(MathError::MulOverflow(21))?
+                .checked_div(pool_token_supply.into())
+                .ok_or(MathError::DivByZero(13))?
+                .try_into()
+                .map_err(|_| MathError::CastOverflow(6))?,
+        );
+        new_reserves.push(
+            old_reserves[i]
+                .checked_add(*amounts.last().unwrap())
+                .ok_or(MathError::SubUnderflow(23))?,
+        );
+    }
+    Ok((amounts, new_reserves))
```

Comment:
`new_reserves` is not used anywhere but in the tests which themselves assert that we add `amount` to `old_reserves` but those `new_reserves` are not used anywhere. We can remove `new_reserves` entirely here and then `compute_deposit_*` and `compute_withdraw_*` are equivalent.

Reply:
[Remove duplicate method](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/5/commits/548baf8781393cb2706b65fb3bd1451c0c4d539f)